components/event-bus/cmd/mesh-namespaces-migration/subscription.go

```diff
@@ -0,0 +1,236 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	kneventingv1alpha1 "knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	kneventingclientset "knative.dev/eventing/pkg/client/clientset/versioned"
+	"knative.dev/pkg/apis"
+	knv1alpha1 "knative.dev/pkg/apis/v1alpha1"
+
+	kymaeventingv1alpha1 "github.com/kyma-project/kyma/components/event-bus/apis/eventing/v1alpha1"
+	kymaeventingclientset "github.com/kyma-project/kyma/components/event-bus/client/generated/clientset/internalclientset"
+)
+
+// CloudEvents attributes
+const (
+	sourceAttr           = "source"
+	eventTypeAttr        = "type"
+	eventTypeVersionAttr = "eventtypeversion"
+)
+
+const (
+	defaultBrokerName = "default"
+	brokerLabel       = "eventing.knative.dev/broker"
+)
+
+type triggersByNamespaceMap map[string][]kneventingv1alpha1.Trigger
+type subscriptionsList []kymaeventingv1alpha1.Subscription
+
+// subscriptionMigrator performs migrations of Kyma Subscriptions to Knative Triggers.
+type subscriptionMigrator struct {
+	kymaClient    kymaeventingclientset.Interface
+	knativeClient kneventingclientset.Interface
+
+	subscriptions       subscriptionsList
+	triggersByNamespace triggersByNamespaceMap
+}
+
+// newSubscriptionMigrator creates and initializes a subscriptionMigrator.
+func newSubscriptionMigrator(kymaClient kymaeventingclientset.Interface,
+	knativeClient kneventingclientset.Interface, namespaces []string) (*subscriptionMigrator, error) {
+
+	m := &subscriptionMigrator{
+		kymaClient:    kymaClient,
+		knativeClient: knativeClient,
+	}
+
+	if err := m.populateSubscriptions(namespaces); err != nil {
+		return nil, err
+	}
+	if err := m.populateTriggers(namespaces); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// populateSubscription populates the local list of Kyma Subscriptions.
+func (m *subscriptionMigrator) populateSubscriptions(namespaces []string) error {
+	var kymaSubscriptions []kymaeventingv1alpha1.Subscription
+
+	for _, ns := range namespaces {
+		subs, err := m.kymaClient.EventingV1alpha1().Subscriptions(ns).List(metav1.ListOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "listing Subscriptions in namespace %s", ns)
+		}
+		kymaSubscriptions = append(kymaSubscriptions, subs.Items...)
+	}
+
+	m.subscriptions = kymaSubscriptions
+
+	return nil
+}
+
+// populateTriggers populates the local triggersByNamespace map.
+func (m *subscriptionMigrator) populateTriggers(namespaces []string) error {
+	triggersByNamespace := make(triggersByNamespaceMap)
+
+	for _, ns := range namespaces {
```

Comment:
I'd really recommend using SharedInformerFactory.Lister() to get knative resources, to avoid nested loops.

Reply:
The example you mention is in the context of a long running program. An Informer is not what you would want to use here. See previous comment.